artichoke-backend/src/extn/core/time/args.rs

```diff
@@ -21,134 +21,133 @@ impl Default for Args {
             hour: 0,
             minute: 0,
             second: 0,
-            micros: 0,
+            nanoseconds: 0,
         }
     }
 }
 
-impl Args {
-    pub fn year(&self) -> Result<i32, Error> {
-        i32::try_from(self.year).map_err(|_| ArgumentError::with_message("year out of range").into())
-    }
+impl TryConvertMut<&[Value], Args> for Artichoke {
+    type Error = Error;
 
-    pub fn month(&self) -> Result<u8, Error> {
-        match u8::try_from(self.month) {
-            Ok(month @ 1..=12) => Ok(month),
-            _ => Err(ArgumentError::with_message("mon out of range").into()),
+    fn try_convert_mut(&mut self, args: &[Value]) -> Result<Args, Self::Error> {
+        // Time args should have a length of 1..=8 or 10. The error does not
+        // give a hint that the 10 arg variant is supported however (this is
+        // the same in MRI).
+        if let 0 | 9 | 11 = args.len() {
+            let mut message = br#"wrong number of arguments (given "#.to_vec();
+            message.extend(args.len().to_string().bytes());
+            message.extend_from_slice(b", expected 1..8)");
+            return Err(ArgumentError::from(message).into());
         }
-    }
 
-    pub fn day(&self) -> Result<u8, Error> {
-        match u8::try_from(self.day) {
-            Ok(day @ 1..=31) => Ok(day),
-            _ => Err(ArgumentError::with_message("mday out of range").into()),
-        }
-    }
+        let mut args = args.to_vec();
```

Comment:
Note, we need to change this to an owned structure because we're mutating it in the 10 argument variant. I'm sure there's a way to do this just with slices but couldn't figure it out. Alternative is to just make the input `mut` :shrug:

Reply:
I'd rather us take a `&mut [Value]` for the `TryConvertMut` impl and push the allocation to callers if they need to. The API design principle here is _ask for what you need_ and indeed, callers of this converter already have access to a mutable slice, so we can save an allocation.

Reply:
Yes, good to know! It looks a whole lot better than a random new allocation hanging around.